aclocal.m4

```diff
@@ -501,33 +501,45 @@ AC_DEFUN([OCAML_QUOTED_STRING_ID], [
   done
 ])
 
-AC_DEFUN([OCAML_CC_SUPPORTS_ATOMIC], [
-  OCAML_CC_SAVE_VARIABLES
-
-  opts=""
-  AS_IF([test -n "$1"],[CFLAGS="$CFLAGS $1"; opts="$1"])
-  AS_IF([test -n "$2"],[LIBS="$LIBS $2"; opts="${opts:+$opts }$2"])
-  AC_MSG_CHECKING(m4_normalize([if $CC supports _Atomic types with
-    ${opts:-no additional options}]))
-
-  AC_LINK_IFELSE([AC_LANG_PROGRAM([[
+# OCAML_CC_C11_ATOMIC_CFLAGS([LIBS])
+AC_DEFUN([OCAML_CC_C11_ATOMIC_CFLAGS], [
+  AC_CACHE_CHECK([for options needed to enable C11 atomic support],
+    [ocaml_cv_prog_cc_c11_atomic_cflags],
+    [ocaml_cv_prog_cc_c11_atomic_cflags='cannot detect'
+    for ocaml_arg in dnl
+        ''dnl
+        '-experimental:c11atomics'dnl
+        '-std:c11'dnl remove with Autoconf 2.73 as cl will default to C11
+        '-std:c11 -experimental:c11atomics'dnl same
+    ; do
+      OCAML_CC_SAVE_VARIABLES
+      CFLAGS="$CFLAGS $ocaml_arg"
+      AS_IF([test -n "$1"], [LIBS="$LIBS $1"])
+      AC_LINK_IFELSE([AC_LANG_PROGRAM([[
 #include <stdint.h>
 #include <stdatomic.h>
-    ]],[[
+      ]],[[
   _Atomic int64_t n;
   int m;
   int * _Atomic p = &m;
   atomic_store_explicit(&n, 123, memory_order_release);
   * atomic_exchange(&p, 0) = 45;
   if (atomic_load_explicit(&n, memory_order_acquire))
     return 1;
-  ]])],
-  [cc_supports_atomic=true
-   AC_MSG_RESULT([yes])],
-  [cc_supports_atomic=false
-   AC_MSG_RESULT([no])])
-
-  OCAML_CC_RESTORE_VARIABLES
+      ]])],
+      [AS_IF([test x"$ocaml_arg" = x],
+        [ocaml_cv_prog_cc_c11_atomic_cflags='none needed'],
+        [ocaml_cv_prog_cc_c11_atomic_cflags="$ocaml_arg"])
+      OCAML_CC_RESTORE_VARIABLES
+      break])
+      OCAML_CC_RESTORE_VARIABLES
+    done])
+  AS_CASE([$ocaml_cv_prog_cc_c11_atomic_cflags],
+    ['cannot detect'],
+      [AC_MSG_FAILURE(m4_normalize([C11 atomic support is required, use another
+        C compiler]))],
+    ['none needed'], [:],
```

Comment:
These magic strings worry me a bit - `error` and `none` (that way they look like magic constants - and we know they can't be the required flags) and less a message which someone might be tempted to change to "unable to detect" and "none required" without realising they matter!